src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,694 @@
+        // types before performing lookups in this dictionary.
+        private static readonly Dictionary<Type, QueryParameterFormatter> _queryParameterFormatters = new()
+        {
+            [typeof(string)] = value => (string)value,
+            [typeof(bool)] = value => Format((bool)value),
+            [typeof(DateTime)] = value => Format((DateTime)value),
+            [typeof(decimal)] = value => Format((decimal)value),
+            [typeof(double)] = value => Format((double)value),
+            [typeof(float)] = value => Format((float)value),
+            [typeof(Guid)] = value => Format((Guid)value),
+            [typeof(int)] = value => Format((int)value),
+            [typeof(long)] = value => Format((long)value),
+        };
+
+        private static string Format(bool value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(bool? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(DateTime value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(DateTime? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(decimal value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(decimal? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(double value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(double? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(float value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(float? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(Guid value)
+            => value.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string? Format(Guid? value)
+            => value?.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string Format(int value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(int? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(long value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(long? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private struct QueryStringBuilder
+        {
+            private readonly StringBuilder _builder;
+
+            private bool _hasNewParameters;
+
+            public string UriWithQueryString => _builder.ToString();
+
+            public QueryStringBuilder(ReadOnlySpan<char> uriWithoutQueryString)
+            {
+                _builder = new();
```

Comment:
Super-minor suggestion - please ignore if inconvenient! If we happened to know the length of the original string, we could initialize the `StringBuilder` with that capacity (or maybe that capacity + 20 chars or something). Otherwise the default capacity is only 16 chars after which it has to do more internal copying or other housekeeping.